### ISSUE TITLE
`https://gorouter.dev/` points to `pub.dev`

### ISSUE THREAD

**Issue by iLoveDocs** [flutter/flutter#112159]
https://gorouter.dev/ started pointing to the pub package [`go_router`](https://pub.dev/packages/go_router) and all its documentation seems to be lost.

**Comment by exaby73** [comment]
I can see this change currently. Not sure if it's intended or not. Labelling for insights from the team.

**Comment by chunhtai** [comment]
This is expected as we plan to use readme as central place for documentations. The some of the documentation got moved to https://pub.dev/documentation/go_router/latest/ and https://github.com/flutter/packages/tree/main/packages/go_router/example

Feel free to file issues or prs if you think there are parts of documentations that need to be improved.

**Comment by ryanheise** [comment]
> we plan to use readme as central place for documentation.

There is so much more in https://gorouter.dev/ than would ever be appropriate to fit in a README, including a much more slow-paced introduction and tutorial for each concept, in-depth discussion of advanced scenarios and topics, and translations of the whole documentation into Japanese. pub.dev doesn't any way to publish different translations of the README (that I'm aware of). If we take a word count of the current README, it is 728 words. If we take a word count of the full documentation on the website, it is 13,703 words - a generous 19x larger. But inappropriate for a README.

> This is expected

But look at the comments here - it's apparently not expected by anyone else.

If there is a plan to eventually bring back this valuable resource at a new location, or to rewrite everything that used to be in the old resource, why not at least provide a link to the old documentation UNTIL it has been migrated to the new location?

It would be quite challenging for developers to learn how to use some of the advanced aspects of go_router without the old guides.

 > Feel free to file issues or prs if you think there are parts of documentation that need to be improved.

Why encourage people to create new PRs from scratch when all of that documentation had already existed in the past? It will lead to people unnecessarily duplicating effort.

For anyone stumbling upon this, you can find the source files for the original documentation here:

https://github.com/csells/go_router/tree/main/docs

And the Japanese translation is still browsable at this site:

https://zenn.dev/inari_sushio/scraps/01ef7604a4b934

**Comment by iLoveDocs** [comment]
@ryanheise Thanks so much for saying my heart out in detail. 

>  If we take a word count of the current README, it is 728 words. If we take a word count of the full documentation on the website, it is 13,703 words - a generous 19x larger.

> Why encourage people to create new PRs from scratch when all of that documentation had already existed in the past? It will lead to people unnecessarily duplicating effort.

Can't agree more!

**Comment by exaby73** [comment]
@ryanheise @iLoveDocs You can find an archived version of the docs [here](https://twitter.com/biz84/status/1573243276257263616?s=21&t=H5YNnonLtZCd0P4S2iBpQw)

**Comment by chunhtai** [comment]
Thanks for the reply. Let me explain a few things.

### Why we do this

The some of the old guides were outdated or incorrect due to API changes in the new go_router version. while some part was still useful, the other part may lead to undesired behavior. Flutter does not own the `https://gorouter.dev/` so it takes extra effort to contact the owner and update docs. At some point we would have to migrate the doc to somewhere owned by flutter, so we decided Readme with link to the API doc or example is a good place to put them, and it also matches the style of other first party packages.

### Parity between old docs and new docs.

besides the out-dated part of old docs, there are still some parity between old and new docs. some of them were moved to the API doc directly, for example https://pub.dev/documentation/go_router/latest/go_router/GoRoute/redirect.html. Some of was removed by purposes and some of them was straight up missing due to oversight. If the latter two cause confusions, I sincerely apologize. Either we made a mistake and think something that should be removed but shouldn't or we missed it entirely. We are still trying our best to improve the documentation. If you think some part or some APIs needs more explanation or mentions from the readme, please file an issue to discuss that so we can prioritize it.

**Comment by ryanheise** [comment]
Thank you for the reply, but I found it to be quite a confusing reply.

> Why we do this: Some of the old guides were outdated or incorrect due to API changes in the new go_router version.

A decision to not continue to maintain the old tutorials/guides should not be disguised as a necessary result of the old tutorials/guides being out of date. The reason they are out of date is because of an explicit choice to not maintain them. It is Flutter's right to make that choice, but it would be more transparent to address the feedback by saying "we decided not to keep the old tutorials/guides up to date for reasons X, Y, Z".

> Flutter does not own the https://gorouter.dev/ so it takes extra efforts to contact the owner and update docs.

The domain ownership is not a blocker. Flutter owns https://docs.flutter.dev/ and could have maintained the new copy of the guides there before letting them get out of date.

> We are still trying our best to improve the documentation. If you think some part or some APIs needs more explanation or mentions from the readme, please file an issue to discuss that so we can prioritize it.

No amount of improvements to the README will address my concern, because if the guide/tutorial-like nature of gorouter.dev were to be incorporated into the README, then the README page would become inappropriately large. And updating the API documentation also won't address my concerns because API documentation pages are also inappropriate places to host guides and tutorials. Guides and tutorials are based on use cases and goals that have a broader scope than just a single class, method, field or function API (which is the scope of an API doc). One example would be nested navigation. This is a large topic, too big for the README. But then it is also a topic that you can't just spread out into the individual API doc pages because what's needed is a guide showing how all the pieces fit together, not just separate pages about each individual API. The API page on ShellRoute just shows an example for the ShellRoute class but does not give a wholistic example showing how this can fit with bottom navigation. And there are many other scenarios missing such as how to handle a tab view which might be handled differently. And there is also the scenario of wanting to update the AppBar or access the drawer from nested routes. A developer cannot easily find the answers to these questions by simply reading through the README or reading through the API, they need a guide.

I would expect the README to be reserved for a summary of usage, the API docs to be reserved for docs for the respective APIs, and tutorials/guides to be hosted elsewhere, such as docs.flutter.dev.

Don't get me wrong, it's nice to have a good quality README, and some people will be able to get the gist of things from the README alone, but many people will be lost without a guide. Navigation is a complex topic deserving of such a guide.

> it also matches the style of other first party packages.

I don't think all packages are equal. If the usage of a package can be easily explained in a README alone, then great! But that is not the case for packages that have greater complexity. For example, this is the case for the various firebase packages which have necessary guides linked and hosted at another more appropriate place.

**Comment by chunhtai** [comment]
> it would be more transparent to address the feedback

We do have an issue that tracks this efforts https://github.com/flutter/flutter/issues/102171
Although I agree we could have done better communicating to the public.

> The domain ownership is not a blocker. Flutter owns https://docs.flutter.dev/ and could have maintained the new copy of the guides there before letting them get out of date.

The domain was already outdated before we start the migration, so some works were needed before the migration. So it downed to where was the best place to do these, readme or a section in `https://docs.flutter.dev/`. I chose the earlier for easier maintenance, and I think we can mimic the original gorouter.dev to have a lot of section and link to different page in the readme too. It was just that we decided to separate some doc to API page. We can always change that back to be a more goroute.dev style if the discussion turned out to favor that in the end.

> And updating the API documentation also won't address my concerns because API documentation pages are also inappropriate places to host guides and tutorials

Although I slightly disagree the API documentation pages are inappropriate, I think improving doc will not be bound to readme or API page update. Depends on the topics, we may choose to write a doc in https://docs.flutter.dev/,  medium post, or youtube vids. We can be flexible, this is just a starting point.

**Comment by ryanheise** [comment]
> > The domain ownership is not a blocker. Flutter owns https://docs.flutter.dev/ and could have maintained the new copy of the guides there before letting them get out of date.
> 
> The domain was already outdated before we start the migration

That's a bit of an exaggeration. You cite the "Declarative Routing" page as one example, which said that it didn't matter which order you defined routes in. But that was not "out of date" at the time. It was stated that it is an invalid usage to define multiple conflicting routes, and given that, it indeed didn't matter what order the routes were defined in. go_router has since been changed to allow multiple matches in the list and to pick the first one, so "now" it's out of date, but at the time of the handover, you could not say that the documentation had already fallen out of date. It fell out of date after the migration.

Duration the migration, I can see the approach you have taken, but in the process, you also cut it down to size substantially, removing content, making the sentences more terse and less readable, less flowing and less connected. It is understandable that the document would be less helpful after cutting it down in size dramatically, but let's please have back all those sections that you cut out. 

> I think we can mimic the original gorouter.dev to have a lot of section and link to different page in the readme too.

Yes, but it fits in the README only because you deleted all of those sections/topics:

* deep linking
* url path strategies
* debugging
* transitions
* keys
* async
* neglect
* tab bars

I don't think it's worth the loss. I hope you will bring back all of that great documentation.

> We can always change that back to be a more goroute.dev style if the discussion turned out to favor that in the end.

The early feedback was "a big downgrade from the previous documentation we had", and that is in line with the latest feedback in this issue.

There are also plenty of questions on StackOverflow about how to solve various navigation problems, and other issues submitted here asking for more documentation and examples on how to solve common navigation problems, so it is clear that what we need isn't a terse README, but a more meaty guide that covers common design scenarios, i.e. what gorouter.dev used to have before. A good start would be to bring back the old doc pages for those topics that had been cut out of the migration.

**Comment by Desync-o-tron** [comment]
As a brand new user, I had to go to outdated YouTube videos to learn go_router because the readme was so sparse.

**Comment by cgestes** [comment]
Still using the old documentation also or directly reading the code.

**Comment by johnpryan** [comment]
@ryanheise thanks for the feedback. I don't agree with most of your points, I hope you understand that the original docs were written by a Googler in their spare time and didn't go through the normal review and quality checking process that other docs go through.

We do cover [deep linking](https://docs.flutter.dev/development/ui/navigation/deep-linking) and [url path strategies](https://docs.flutter.dev/development/ui/navigation/url-strategies) on flutter.dev, which will be kept up to date. I agree we should link to these pages like the old docs.

For debugging, transitions, keys, async, and neglect, etc... Transitions are done using [pageBuilder](https://pub.dev/documentation/go_router/latest/go_router/GoRoute/pageBuilder.html), debugging is done using [debugLogDiagnostics](https://pub.dev/documentation/go_router/latest/go_router/GoRouter-class.html), keys should no longer be necessary for nested navigation, so I consider that page out of date. The async page just shows how to use FutureBuilder, so I would recommend users use the [Fetch data from the internet](https://docs.flutter.dev/cookbook/networking/fetch-data) cookbook article on flutter.dev instead. (This page doesn't seem to have anything to do with go_router.)

The tab bars documentation is out of date with the new ShellRoute class (and more out of date still when stateful nested navigation is shipped: https://github.com/flutter/packages/pull/2650 ).

Keep in mind that the [original docs are still available](https://docs.page/csells/go_router), but as @chunhtai pointed out, they are out of date, and we plan to document things in the README, API documentation, or on flutter.dev, not using docs.page going forward.

**Comment by cgestes** [comment]
I think we are just saying the old doc is more useful and complete than what is currently available. Updating it's content and integrating everything that is missing in the current documentation would probably make everyone happy.

some example of hard to understand information: what's the difference between go and push.

**Comment by ryanheise** [comment]
> I don't agree with most of your points

Most of your response was of the form: X is out of date, Y is out of date, Z is out of date, as if I had been claiming otherwise. But I didn't claim otherwise, and that was certainly not my point.

> For debugging, transitions, keys, async, and neglect, etc... Transitions are done using [pageBuilder](https://pub.dev/documentation/go_router/latest/go_router/GoRoute/pageBuilder.html), debugging is done using [debugLogDiagnostics](https://pub.dev/documentation/go_router/latest/go_router/GoRouter-class.html), keys should no longer be necessary for nested navigation, so I consider that page out of date. The async page just shows how to use FutureBuilder, so I would recommend users use the [Fetch data from the internet](https://docs.flutter.dev/cookbook/networking/fetch-data) cookbook article on flutter.dev instead. (This page doesn't seem to have anything to do with go_router.)

We actually agree that some of these pages are now out of date, but I'm saying that the new README, despite as you now say going through a supposedly higher level of quality assurance, is to many people of lower quality than the original documentation because it leaves out all of these well-written topic guides that were so illuminating and helpful to new users of go_router back when those guides were still not out of date, and it is unfortunate that the Flutter team has allowed this more appreciated documentation to become out of date, _seemingly without acknowledging how much they were appreciated_.

If any of these pages are now out of date, it strengthens my point because it means that developers now have no guidance on how to accomplish these omitted topics in the new version of go_router.

> The tab bars documentation is out of date with the new ShellRoute class (and more out of date still when stateful nested navigation is shipped: [flutter/packages#2650](https://github.com/flutter/packages/pull/2650) ).

The reason that this needed its own topic guide is because the TabBarView scenario and the BottomNavigationBar scenario are completely different. In the latter, you would typically use nested navigators to maintain the stack for each page, but in a TabBarView, there is only a single page and it needs to be a stateful widget so that it can animate sideways to reveal the different content. You don't implement these in the same way. One pushes a page onto the stack, the other needs a hook so that you can call the appropriate method on a controller to trigger the sideways animation between tab views.

And the way you keep stressing how it's getting "more out of date" does not help your case, and does not have good optics when the point was that there is dissatisfaction with the Flutter Team's choice to allow these guides to become out of date in the first place, while also strangely not being able to acknowledge how much we appreciated these guides. As I have pointed out in a previous comment, these guides were largely not out of date at the time of handover. ShellRoute, for instance, did not even exist at the time of handover.

>  and we plan to document things in the README, API documentation, or on flutter.dev, not using docs.page going forward.

if you covered these guide topics on flutter.dev, that would be perfect. There is nothing too different between "using docs.page" vs using "flutter.dev" aside from the hosting location, and as I commented above:

> Flutter owns https://docs.flutter.dev/ and could have maintained the new copy of the guides there before letting them get out of date.

I will leave it here, since everything has been said. Hopefully the Flutter team can acknowledge what people liked about the old guides before committing to a plan of action that simply dismisses the old guides as out of date and (ironically) not meeting Google's quality requirements.

**Comment by johnpryan** [comment]
> the Flutter team has allowed this more appreciated documentation to become out of date, seemingly without acknowledging how much they were appreciated.

We don't own the domain nor the keys to publish, and we have no way to use `docs.page` for packages in the flutter/packages repo so we couldn't just copy-paste these and keep them up to date.

> it is unfortunate that the Flutter team has allowed this more appreciated documentation to become out of date, seemingly without acknowledging how much they were appreciated.

What are you using to measure appreciation? We don't _own_ the original gorouter docs so we don't have any insight into the page views.  We can't update these docs, which is why they are out of date.

> there is dissatisfaction with the Flutter Team's choice to allow these guides to become out of date in the first place [...]  these guides were largely not out of date at the time of handover

That's true, but again we can't update the docs so I disagree that this was a choice on our end. If we were given the keys to publish, and got some help setting that up in the flutter/packages repo perhaps things would be different. 

> Flutter owns https://docs.flutter.dev/ and could have maintained the new copy of the guides there before letting them get out of date.

That would require a review from our technical writers. cc: @sfshaza2 

> Hopefully the Flutter team can acknowledge what people liked about the old guides before committing to a plan of action that simply dismisses the old guides as out of date and (ironically) not meeting Google's quality requirements.

Ack.

**Comment by ryanheise** [comment]
> again we can't update the docs, which is why they are out of date.

OK, I think I can see where both you and @chunhtai are misunderstanding. It seems you think that someone is asking you to edit the docs specifically at the gorouter.dev site. But nobody in this issue ever said we must keep the old domain name to keep the docs maintained. The docs could be migrated and maintained elsewhere to keep them up to date. The old README suggested that the documentation would be migrated to a new place where you CAN edit it.

Specifically, the old README made the announcement that the package would be moved into the flutter/packages repository and that the gorouter.dev documentation would be moved to docs.flutter.dev, so that was the expectation. But even if it ended up being moved to the README instead of flutter.dev, it is of course still being migrated to some place where you DO have the ability to update the docs AND keep them up to date. E.g. as a start: `cat docs/*.mdx > README.md`.

The issue, then, is that in the migration process, we lost a lot of content because of the chosen approach to build up the README from scratch. That was always going to be a slow process that would leave us with unhelpful holes in the documentation for a considerable time. It would have been better to start with the old (excellent) documentation, maintain it, and to incrementally mold it to to Google's preferred documentation style over time. This would maximise availability of the documentation during the migration.

Feel free to thumbs down this feedback again, I am just wishing to offer my feedback on this migration but you can do with it what you will.

> > The tab bars documentation is out of date with the new ShellRoute class (and more out of date still when stateful nested navigation is shipped: [flutter/packages#2650](https://github.com/flutter/packages/pull/2650) ).
> 
> The reason that this needed its own topic guide is because the TabBarView scenario and the BottomNavigationBar scenario are completely different. In the latter, you would typically use nested navigators to maintain the stack for each page, but in a TabBarView, there is only a single page and it needs to be a stateful widget so that it can animate sideways to reveal the different content. You don't implement these in the same way. One pushes a page onto the stack, the other needs a hook so that you can call the appropriate method on a controller to trigger the sideways animation between tab views.

Since you have continued to respond, do you have any response to this comment? Wouldn't it have been better to copy the old doc page into the new documentation, and then maintain that documentation as go_router continues to change?

Otherwise, how else are developers going to learn how to implement TabBar navigation inside go_router in the meantime?

You say that this page is out of date now that we have the new ShellRoute class, but this is wrong for the reasons given above. Nested navigators don't even help to solve this problem(*), or at the very least, it's an orthogonal problem. The problem in question here is not how to nest navigators but rather how to animate between the different tab views at the same hierarchical level. You still need some hook to be able to call the appropriate method (or setter) on the controller to trigger the tab switching animation. Any response to that or are you just leaving it at your last response? That is, that you should dismiss the majority of what I'm saying?

(*) Who is using nested navigators in tab bars anyway? Isn't this an anti-pattern? In material design, bottom navigation bars are the ones where we navigate across independent nested hierarchies, while tab bars are where we are navigating across the "same" level in the hierarchy. I am happy to be corrected on that point, though, it does not affect the above argument. The problem of how to use tab bars in go_router specifically relates to what hook you need to respond to and link up with the appropriate controller method that animates the tabs.

> > Hopefully the Flutter team can acknowledge what people liked about the old guides before committing to a plan of action that simply dismisses the old guides as out of date and (ironically) not meeting Google's quality requirements.
> 
> Ack.

Don't be too enthusiastic now :-)

> What are you using to measure appreciation? [...] we don't have any insight into the page views.

For qualitative, I suggest reading the feedback here and in other issues (and the pull request that committed the new README) from the 7+ people who have shared their views.

If you want quantitative, I suggest counting the number of comments and up-votes.

I think the consensus is clear: we love the original, comprehensive docs. They cover topics that are no longer mentioned in the new README, and so it would be a good idea to bring these missing pages back (e.g. as sections in the README, and ideally incorporated into less-cramped guides on flutter.dev).

**Comment by johnpryan** [comment]
> The docs could be migrated and maintained elsewhere to keep them up to date.

I don't think so, the original docs used docs.page, which we don't have tooling to support right now. Adding them to flutter.dev would require a review from our technical writers, but that will take some time since the content is also out of date now.

> the old README made the announcement that the package would be moved into the flutter/packages repository and that the gorouter.dev documentation would be moved to docs.flutter.dev

We chose to focus on adding features for 5.0.0 instead, but adding GoRouter docs to flutter.dev is still important for us.

> The reason that this needed its own topic guide is because the TabBarView scenario and the BottomNavigationBar scenario are completely different. In the latter, you would typically use nested navigators to maintain the stack for each page, but in a TabBarView, there is only a single page and it needs to be a stateful widget so that it can animate sideways to reveal the different content. You don't implement these in the same way. One pushes a page onto the stack, the other needs a hook so that you can call the appropriate method on a controller to trigger the sideways animation between tab views.

This isn't a pattern we intend to recommend, since we found that users have a hard time understanding it. Instead, the sample should use ShellRoute with TabBar, not a TabBarView.

> If you want quantitative, I suggest counting the number of comments and up-votes.

We do, that's why we've been working on other issues with more upvotes.

> I think the consensus is clear: we love the original, comprehensive docs. They cover topics that are no longer mentioned in the new README, and so it would be a good idea to bring these missing pages back (e.g. as sections in the README, and ideally incorporated into less-cramped guides on flutter.dev).

The README is probably the best place for these right now, but we plan to add more to flutter.dev also. Feel free to send us a pull request if it's urgent.
